crates/pathfinder/src/state/sync/l2.rs

```diff
@@ -637,6 +623,375 @@ async fn download_block(
+                    chain,
+                    chain_id,
+                    block_validation_mode,
+                )
+                .and_then(
+                    |(
+                        transaction_commitment,
+                        event_commitment,
+                        receipt_commitment,
+                        state_diff_commitment,
+                    )| {
+                        verify_signature(
+                            block.block_hash,
+                            state_diff_commitment,
+                            &signature,
+                            sequencer_public_key,
+                            BlockValidationMode::AllowMismatch,
+                        )
+                        .map_err(|err| err.into())
+                        .map(|_| {
+                            (
+                                block,
+                                state_update,
+                                signature,
+                                transaction_commitment,
+                                event_commitment,
+                                receipt_commitment,
+                                state_diff_commitment,
+                            )
+                        })
+                    },
+                );
+
+                let t_verification = t_verification.elapsed();
+                tracing::trace!(elapsed=?t_verification, "Block verification done");
+
+                let _ = tx.send(result);
+            });
+
+            let (
+                block,
+                state_update,
+                signature,
+                transaction_commitment,
+                event_commitment,
+                receipt_commitment,
+                state_diff_commitment,
+            ) = rx
+                .await
+                .expect("Panic on rayon thread while verifying block")
+                .context("Verifying block contents")?;
+
+            let t_declare = std::time::Instant::now();
+            let downloaded_classes = download_new_classes(&state_update, &sequencer, storage)
+                .await
+                .with_context(|| {
+                    format!("Handling newly declared classes for block {block_number:?}")
+                })?;
+            let t_declare = t_declare.elapsed();
+
+            let timings = Timings {
+                block_download: t_block,
+                class_declaration: t_declare,
+                signature_download: t_signature,
+            };
+
+            Ok::<_, anyhow::Error>((
+                block,
+                state_update,
+                signature,
+                transaction_commitment,
+                event_commitment,
+                receipt_commitment,
+                state_diff_commitment,
+                downloaded_classes,
+                timings,
+            ))
+        }
+        .in_current_span()
+    });
+    let mut stream = futures::stream::iter(futures).buffered(fetch_concurrency.get());
```

Comment:
I understand we need to maintain order here for the emits to maintain the order too. But this means that we unnecessarily wait for the entire batch to "fill up". How about using `buffer_unordered` and maintaining a queue with "gaps" that'd be filled with lagging blocks. Such queue would then be able `pop_front` until it reaches a gap at which point it should wait for the gap to be filled in.
Pros:
- downloading is not artificially halted and happens asap
Cons:
- event consumption requires proper block order, so no gain here

Reply:
Continuing the above "loose" thought: maybe we could do "trie"-ed insertions in order and anything else out of order :thinking: .

Ultimately I guess the speedup is so significant right now that we don't really need to fine tune it anymore without getting into needless complexity :bow: .